### PR TITLE
Standardize batching filters to read routing from message (Signalfxbatch)

### PR DESCRIPTION
**Pull request by nathanleiby**
JIRA: [INFRA-1959](https://clever.atlassian.net/browse/INFRA-1959)

This adds support for the signalfx batching filter to read all routing information from the message fields themselves, instead of the setting it in a config object.

To configure new routing, you'll simply pass new information on the message itself, rather than redeploying Heka with an updated config.

So far 

- This only updates the signalfx batching filter (influx still needs to be done) -- UPDATE: will do influx in another PR
- This updates the filter inline (we probably want to fork it and keep the old one around, while migrating) -- UPDATE: forked it and kept old one
- I added some Lua mocks/tests (I'll want to refactor these when building the influx filter, as well as move them somewhere that makes sense) -- UPDATE: used [busted](https://github.com/Olivine-Labs/busted), a lua test framework

### CODE REVIEW

**Comment by nathanleiby** [comment]
@xavi- If you have feedback so far, I'd welcome it!

File: lua/filters/test_signalfxbatch.lua [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+assert(shallow_compare(decoded["counter"][1]["dimensions"], expected_dimensions))
+
+-- Process 2 messages -- does it batch?
+process_result = process_message()
+assert(process_result == 0, "Should process_message successfuly")
+process_result = process_message()
+assert(process_result == 0, "Should process_message successfuly")
+flush()
+injected = injected_payloads()
+assert(#injected == 2, "There should be three messages flushed")
```
Comment (nathanleiby) [inline lua/filters/test_signalfxbatch.lua]: should be 2 .. batch 1 (1 counters), batch 2 (2 counters)

File: lua/filters/signalfxbatch.lua [hunk 2]
```diff
@@ -2,98 +2,105 @@
...
+    message_matcher = "TRUE"
     type = "SandboxFilter"
     script_type = "lua"
-    filename = "lua/filters/signalfxbatch.lua"
-       [SignalfxBatchFilter.config]
-       metric_name = "test-metric.%{title}.%{Hostname}"
-       value_field = "metric_value"
-       dimensions = "Hostname Severity"
-       max_count = 5
+    filename = "lua/filters/signalfxbatch_messsage.lua"
```
Comment (nathanleiby) [inline lua/filters/signalfxbatch.lua]: fix filename

File: lua/filters/mocks.lua [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+-- return all injected payloads
+function injected_payloads()
+    -- Debug
+    debug("injected_payloads")
+    for i, v in ipairs(MOCKS.injected_payloads) do debug(i .. "\t" .. v) end
+    return MOCKS.injected_payloads
+end
+
+function shallow_compare(t1, t2, max_recursion)
```
Comment (xavi-) [inline lua/filters/mocks.lua]: Should `max_recursion` be removed?

File: lua/filters/signalfxbatch.lua [hunk 4]
```diff
@@ -2,98 +2,105 @@
...
 
+- ticker_interval (int, optional, defaults to never sending a ticker event)
+    Max time delay before a batch is flushed from the filter.
 
-- max_count (int, optional, defaults to 20)
-    Max number of messages before a batch is flushed from the filter.
+- value_field (string, required)
+    The `fieldname` to use as the value for the metric in signalfx. If the `value`
+    field is not present this encoder will set one as the value for counters: `1`.
+    A value of `0` will be used for `gauges`.
```
Comment (xavi-) [inline lua/filters/signalfxbatch.lua]: Worth documenting `stat_type`?  Also smallest of nit, is there another name for `stat_type`?

Comment (nathanleiby) [inline lua/filters/signalfxbatch.lua]: maybe "metric_type"? what were you thinking?

Comment (nathanleiby) [inline lua/filters/signalfxbatch.lua]: https://prometheus.io/docs/concepts/metric_types/

**Pull request merged**